binary_search_tree/tree.py

```diff
@@ -14,42 +17,131 @@ class Tree:
+        if current_node == None:
+            return TreeNode(key, value)
+        if key <= current_node.key:
+            current_node.left = self.add_helper(current_node.left, key, value)
+        else:
+            current_node.right = self.add_helper(current_node.right, key, value)
+        return current_node
+
+
+    # Time Complexity:O(log n) 
+    # Space Complexity:O(1)
     def add(self, key, value = None):
-        pass
+        if self.root == None:
+            self.root = TreeNode(key, value)
+        else:
+            self.add_helper(self.root, key, value)
+
+
 
-    # Time Complexity: 
-    # Space Complexity: 
+    # Time Complexity: O(log n) 
+    # Space Complexity: O(1)
     def find(self, key):
-        pass
+        if self.root == None:
+            return None
+        
+        current = self.root
+        while current != None:
+            if current.key == key:
+                return current.value
+            elif current.key < key:
+                current = current.right
+            else:
+                current = current.left
+        return None
+    def inorder_helper_f(self,current_node, in_order_list):
+        if current_node == None:
+            return in_order_list
+        
+        self.inorder_helper_f(current_node.left, in_order_list)
+        in_order_list.append({"key": current_node.key,"value":current_node.value})
+        self.inorder_helper_f(current_node.right, in_order_list)
+        return in_order_list
 
-    # Time Complexity: 
-    # Space Complexity: 
+    # Time Complexity: O(log n) 
+    # Space Complexity: O(1)
     def inorder(self):
-        pass
+        in_order_list = []
+        
+        return self.inorder_helper_f(self.root,in_order_list)
+        
+
+    def preorder_helper_f(self,current_node,preorder_list):
+        if current_node == None:
+            return  preorder_list
+        
+        preorder_list.append({"key": current_node.key,"value":current_node.value})
 
-    # Time Complexity: 
-    # Space Complexity:     
+        self.preorder_helper_f(current_node.left, preorder_list)
+        self.preorder_helper_f(current_node.right, preorder_list)
+        return preorder_list
+
+
+    # Time Complexity:O(log n) 
+    # Space Complexity: O(1)   
     def preorder(self):
-        pass
+        preorder_list = []
+
+        return self.preorder_helper_f(self.root,preorder_list)
 
-    # Time Complexity: 
-    # Space Complexity:     
+    def postorder_helper_f(self,current_node,  post_order_list):
+        if current_node == None:
+            return  post_order_list
+
+        self.postorder_helper_f(current_node.left, post_order_list)
+        self.postorder_helper_f(current_node.right, post_order_list)
+
+        post_order_list.append({"key": current_node.key,"value":current_node.value})
+        
+        return post_order_list
+    # Time Complexity: O(log n)
+    # Space Complexity: O(1)     
     def postorder(self):
```

Comment:
👍 The space complexity here is O(n) since we are creating a list of all the node values.